src/peft/utils/corda_utils.py

```diff
@@ -0,0 +1,314 @@
+
+        # covariance = input.t() @ input ## (dim, dim)
+        if torch.isnan(input).any():
+            raise Exception("nan in input, break")
+        if torch.isinf(input).any():
+            raise Exception("inf in input, break")
+
+        covariance = input.t().matmul(input)
+        if torch.isnan(covariance).any():
+            raise Exception("nan in covariance, break")
+        if torch.isinf(covariance).any():
+            raise Exception("inf in covariance, break")
+
+        # calculate mean and std
+        mean = input.mean(0)
+        std = input.std(0)
+
+        # add to module
+        module.covariance_matrix += covariance / sample_count
+        module.mean += mean / sample_count
+        module.std += std / sample_count
+
+        # module.covariance_matrix = (module.covariance_matrix + covariance) / 2
+        del covariance, input
+
+    for name, module in target_modules(hooked_model, config):
+        module.covariance_matrix = 0
+        module.mean = 0
+        module.std = 0
+        module.register_forward_hook(hook)
+
+    run_model()
+
+    if hooked_model is not model:
+        targets = {}
+        for name, module in target_modules(model, config):
+            targets[name] = module
+        for name, module in target_modules(hooked_model, config):
+            # There can be modules used only in inference, but not training
+            # Exclude modules not in target model to prevent KeyError in this case
+            if name in targets:
+                targets[name].covariance_matrix = module.covariance_matrix
+                targets[name].mean = module.mean
+                targets[name].std = module.std
+
+    # Save covariance to disk
+    if covariance_file is not None:
+        all_covariance_matrix = {}
+        for name, module in target_modules(model, config):
+            all_covariance_matrix[name] = module.covariance_matrix
+        os.makedirs(os.path.dirname(covariance_file), exist_ok=True)
+        torch.save(all_covariance_matrix, covariance_file)
+
+
+@torch.no_grad()
+def collect_eigens(
+    model: nn.Module,
+    config: LoraConfig,
+):
+    """Call collect_eigens_for_layer and store result in key `eigens` of each layer."""
+    linear_modules = []
+    for name, module in target_modules(model, config):
+        linear_modules.append((name, module))
+    for name, module in tqdm(linear_modules, desc="Collecting eigens"):
+        module.eigens = collect_eigens_for_layer(module, config)
+
+
+@torch.no_grad()
+def collect_eigens_for_layer(
+    linear: nn.Linear,
+    config: LoraConfig,
+) -> CordaEigens:
+    w = linear.weight.data.float()
+    out_dim = w.size(0)
+    in_dim = w.size(1)
+    min_dim = min(in_dim, out_dim)
+
+    if not hasattr(linear, "covariance_matrix"):
+        raise ValueError("build covariance matrix with calib_cov_distribution first")
+    covariance_matrix = linear.covariance_matrix.float()
+
+    damp = 0.01
+    while True:
```

Comment:
Yes, this is used when covariance matrix is not invertible to add compensation on its diagonal until it is invertible to ensure the same inference as the original model before fine-tuning. 

When a matrix is square and invertible, using `pinv` returns nearly the same inverse result as `inv`, but may have larger numerical error than `inv`. When a square  matrix is not invertible, `pinv` and `inv` return different results. Currently `inv` works well in our experiments.  Can we leave it as it is for now, because we need more investigation to see the impact of using `pinv` on the final adaptation performance?

Reply:
Yes, of course you can leave it as is. I was just curious if you had checked if it works.